ui/src/main/java/de/ipb_halle/lbac/material/composition/CompositionHistoryController.java

```diff
@@ -0,0 +1,156 @@
+ *
+ */
+package de.ipb_halle.lbac.material.composition;
+
+import de.ipb_halle.lbac.admission.ACListService;
+import de.ipb_halle.lbac.admission.ACPermission;
+import de.ipb_halle.lbac.admission.GlobalAdmissionContext;
+import de.ipb_halle.lbac.admission.User;
+import de.ipb_halle.lbac.material.Material;
+import de.ipb_halle.lbac.material.common.bean.MaterialBean;
+import de.ipb_halle.lbac.material.common.history.HistoryController;
+import de.ipb_halle.lbac.material.common.service.MaterialService;
+import de.ipb_halle.lbac.material.inaccessible.InaccessibleMaterial;
+import de.ipb_halle.lbac.util.Unit;
+import java.util.Objects;
+
+/**
+ *
+ * @author fmauz
+ */
+public class CompositionHistoryController implements HistoryController<CompositionDifference> {
+
+    MaterialCompositionBean compositionBean;
+    User user;
+    ACListService aclistService;
+    MaterialService materialService;
+    MaterialBean materialBean;
+
+    public CompositionHistoryController(
+            MaterialBean materialBean) {
+        this.compositionBean = materialBean.getCompositionBean();
+        this.user = materialBean.getUserBean().getCurrentAccount();
+        this.aclistService = materialBean.getAcListService();
+        this.materialService = materialBean.getMaterialService();
+    }
+
+    /**
+     *
+     * @param diff
+     */
+    @Override
+    public void applyPositiveDifference(CompositionDifference diff) {
+        for (int i = 0; i < diff.getMaterialIds_new().size(); i++) {
+            Double concentrationNew = diff.getConcentrations_new().get(i);
+            Double concentrationOld = diff.getConcentrations_old().get(i);
+
+            handleDelta(
+                    diff.getMaterialIds_new().get(i),
+                    diff.getMaterialIds_old().get(i),
+                    concentrationOld, concentrationNew,
+                    getUnitFromString(diff.getUnits_old().get(i)),
+                    getUnitFromString(diff.getUnits_new().get(i))
+            );
+
+        }
+    }
+
+    /**
+     *
+     * @param diff
+     */
+    @Override
+    public void applyNegativeDifference(CompositionDifference diff) {
+        for (int i = 0; i < diff.getMaterialIds_old().size(); i++) {
+            Double concentrationOld = diff.getConcentrations_old().get(i);
+            Double concentrationNew = diff.getConcentrations_new().get(i);
+
+            handleDelta(
+                    diff.getMaterialIds_old().get(i),
+                    diff.getMaterialIds_new().get(i),
+                    concentrationNew, concentrationOld,
+                    getUnitFromString(diff.getUnits_new().get(i)),
+                    getUnitFromString(diff.getUnits_old().get(i))
+            );
+        }
+    }
+
+    private Unit getUnitFromString(String unitString) {
+        Unit unit = null;
+        if (unitString != null) {
+            unit = Unit.getUnit(unitString);
+        }
+        return unit;
+    }
+
+    private void handleDelta(
```

Comment:
switch `concentrationToRemove` and `concentrationToInsert` in the method signature